src/test/java/bowlingexercise3/BowlingTest.java

```diff
@@ -61,22 +81,339 @@ public void testBowlingSpares(String turns, int score) {
+
+            "X|-/|X|-/|X|-/|X|-/|X|-/||",
+            "12|-/|X|-/|X|-/|X|-/|X|-/||",
+            "X|34|X|-/|X|-/|X|-/|X|-/||",
+            "X|-/|54|-/|X|-/|X|-/|X|-/||",
+            "X|-/|1/|-/|X|-/|X|-/|X|-/||",
+            "X|-/|X|-/|X|9/|X|-/|X|-/||",
+
+
+    })
+    public void testValidInputStrikeAndSpareNormal10Frames(String turns) {
+
+        Bowling game = new Bowling();
+        Assertions.assertDoesNotThrow(() -> game.turns(turns));
+    }
+
+    @ParameterizedTest
+    @CsvSource(value = {
+
+            "X|X|X|X|X|X|X|X|X|X||XX",
+            "X|X|X|X|X|X|X|X|X|X||X-",
+            "X|X|X|X|X|X|X|X|X|X||X1",
+            "X|X|X|X|X|X|X|X|X|X||X2",
+            "X|X|X|X|X|X|X|X|X|X||X3",
+            "X|X|X|X|X|X|X|X|X|X||X4",
+            "X|X|X|X|X|X|X|X|X|X||X5",
+            "X|X|X|X|X|X|X|X|X|X||X6",
+            "X|X|X|X|X|X|X|X|X|X||X7",
+            "X|X|X|X|X|X|X|X|X|X||X8",
+            "X|X|X|X|X|X|X|X|X|X||X9",
+            "X|X|X|X|X|X|X|X|X|X||1-",
+            "X|X|X|X|X|X|X|X|X|X||2-",
+            "X|X|X|X|X|X|X|X|X|X||3-",
+            "X|X|X|X|X|X|X|X|X|X||4-",
+            "X|X|X|X|X|X|X|X|X|X||5-",
+            "X|X|X|X|X|X|X|X|X|X||6-",
+            "X|X|X|X|X|X|X|X|X|X||7-",
+            "X|X|X|X|X|X|X|X|X|X||8-",
+            "X|X|X|X|X|X|X|X|X|X||9-",
+            "X|X|X|X|X|X|X|X|X|X||1/",
+            "X|X|X|X|X|X|X|X|X|X||2/",
+            "X|X|X|X|X|X|X|X|X|X||3/",
+            "X|X|X|X|X|X|X|X|X|X||4/",
+            "X|X|X|X|X|X|X|X|X|X||5/",
+            "X|X|X|X|X|X|X|X|X|X||6/",
+            "X|X|X|X|X|X|X|X|X|X||7/",
+            "X|X|X|X|X|X|X|X|X|X||8/",
+            "X|X|X|X|X|X|X|X|X|X||9/",
+            "X|X|X|X|X|X|X|X|X|X||--",
+            "X|X|X|X|X|X|X|X|X|X||-/",
+
+
+
+    })
+    public void testValidInputBonusStrike(String turns) {
+
+        Bowling game = new Bowling();
+        Assertions.assertDoesNotThrow(() -> game.turns(turns));
+    }
+
+    @ParameterizedTest
+    @CsvSource(value = {
+
+            "X|X|X|X|X|X|X|X|X|X||1xt",
+            "X|X|X|X|X|X|X|X|X|X||1t",
+            "X|X|X|X|X|X|X|X|X|X||%t",
+            "X|X|X|X|X|X|X|X|X|X|||2",
+            "X|X|X|X|X|X|X|X|X|X||1x",
+            "X|X|X|X|X|X|X|X|X|X||/3",
+            "X|X|X|X|X|X|X|X|X|X||/X",
+            "X|X|X|X|X|X|X|X|X|X||-X",
+
+            "-/|-/|-/|-/|-/|-/|-/|-/|-/|-/||p",
+            "1/|-/|-/|-/|-/|-/|-/|-/|-/|-/||H",
+            "1/|1/|1/|1/|1/|1/|1/|1/|1/|1/||%",
+            "2/|2/|2/|2/|2/|2/|2/|2/|2/|2/||2p",
+            "3/|3/|3/|3/|3/|3/|3/|3/|3/|3/||33",
+            "4/|4/|4/|4/|4/|4/|4/|4/|4/|4/||99",
+            "5/|5/|5/|5/|5/|5/|5/|5/|5/|5/||+|",
+            "6/|6/|6/|6/|6/|6/|6/|6/|6/|6/|||9",
+            "7/|7/|7/|7/|7/|7/|7/|7/|7/|7/||/",
+            "8/|8/|8/|8/|8/|8/|8/|8/|8/|8/||£",
+            "9/|9/|9/|9/|9/|9/|9/|9/|9/|9/||:",
+            "1/|1/|3/|4/|5/|6/|7/|7/|8/|9/||X4",
+
+
+    })
+    public void testInvalidInputBonusFrames(String turns) {
```

Comment:
Maybe this could be split into two groups for invalid bonus strikes and invalid bonus spares? Though I can see why only one test is necessary, it could make it clearer for regression purposes as we know there are two different supporting regular expressions implementing those cases.